docs/guide/migration.md

```diff
@@ -32,6 +32,33 @@ For other projects, there are a few general approaches:
 
 See the [troubleshooting guide](/guide/troubleshooting.html#vite-cjs-node-api-deprecated) for more information.
 
+## Rework `define` and `import.meta.env.*` replacement strategy
+
+In Vite 4, the `define` and `import.meta.env.*` features use different replacement strategies in dev and build:
+
+- In dev, both features are injected as global variables to `globalThis` and `import.meta` respectively.
+- In build, both features are statically replaced with a regex.
+
+This results in a dev and build inconsistency when trying to access the variables, and sometimes even caused failed builds. For example:
+
+```js
+const data = { __APP_VERSION__ }
+// dev: { __APP_VERSION__: "1.0.0" } ✅
+// build: { __APP_VERSION__: undefined } ❌
+
+const docs = 'I like import.meta.env'
+// dev: "I like import.meta.env" ✅
+// build: "I like ({ ... })" ❌
+```
```

Comment:
Maybe in a comment, the `define` option used could be added to help understand these examples? Wouldn't the first one replace as `{ "1.0.0" } 🤔

Reply:
Ah true. Let me check that and update the explanation